samples/matter/common/src/persistent_storage_util.cpp

```diff
@@ -0,0 +1,192 @@
+
+	/* name != nullptr -> If requested key X is empty, process the next one
+		 name != '\0' -> process just node X and ignore all its descendants: X */
+	if (name != nullptr && *name != '\0') {
+		return 0;
+	}
+
+	if (!entry.destination || 0 == entry.destinationBufferSize) {
+		/* Just found the key, do not try to read value */
+		entry.result = true;
+		return 1;
+	}
+
+	uint8_t emptyValue[kEmptyValueSize];
+
+	if (entrySize == kEmptyValueSize && readCb(cbArg, emptyValue, kEmptyValueSize) == kEmptyValueSize &&
+	    memcmp(emptyValue, kEmptyValue, kEmptyValueSize) == 0) {
+		/* There are wrong bytes stored - the same as the magic ones defined above */
+		entry.result = false;
+		return 1;
+	}
+
+	const ssize_t bytesRead = readCb(cbArg, entry.destination, entry.destinationBufferSize);
+	entry.readSize = bytesRead > 0 ? bytesRead : 0;
+
+	if (entrySize > entry.destinationBufferSize) {
+		entry.result = false;
+	} else {
+		entry.result = bytesRead > 0;
+	}
+
+	return 1;
+}
+} /* namespace */
+
+bool PersistentStorage::Init()
+{
+	if (0 != settings_load()) {
+		return false;
+	}
+
+	return true;
+}
+
+bool PersistentStorage::Store(PersistentStorageNode *node, const void *data, size_t dataSize)
+{
+	if (!data || !node) {
+		return false;
+	}
+
+	char key[PersistentStorageNode::kMaxKeyNameLength];
+
+	if (!node->GetKey(key)) {
+		return false;
+	}
+
+	return (0 == settings_save_one(key, data, dataSize));
+}
+
+bool PersistentStorage::Load(PersistentStorageNode *node, void *data, size_t dataMaxSize, size_t &outSize)
+{
+	if (!data || !node) {
+		return false;
+	}
+
+	char key[PersistentStorageNode::kMaxKeyNameLength];
+
+	if (!node->GetKey(key)) {
+		return false;
+	}
+
+	size_t resultSize;
+
+	bool result = LoadEntry(key, data, dataMaxSize, &resultSize);
+
+	if (result) {
+		outSize = resultSize;
+	}
+
+	return result;
+}
+
+bool PersistentStorage::HasEntry(PersistentStorageNode *node)
+{
+	if (!node) {
+		return false;
+	}
+
+	char key[PersistentStorageNode::kMaxKeyNameLength];
+
+	if (!node->GetKey(key)) {
+		return false;
+	}
+
+	return LoadEntry(key);
+}
+
+bool PersistentStorage::Remove(PersistentStorageNode *node)
+{
+	if (!node) {
+		return false;
+	}
+
+	char key[PersistentStorageNode::kMaxKeyNameLength];
+
+	if (!node->GetKey(key)) {
+		return false;
+	}
+
+	if (!LoadEntry(key)) {
+		return false;
+	}
+
+	settings_delete(key);
+
+	return true;
+}
+
+bool PersistentStorage::LoadEntry(const char *key, void *data, size_t dataMaxSize, size_t *outSize)
+{
+	ReadEntry entry{ data, dataMaxSize, 0, false };
+	settings_load_subtree_direct(key, LoadEntryCallback, &entry);
+
+	if (!entry.result) {
+		return false;
+	}
+
+	if (outSize != nullptr) {
+		*outSize = entry.readSize;
+	}
+
+	return true;
+}
+
+bool PersistentStorageNode::GetKey(char *key)
+{
+	if (!key || mKeyName[0] == '\0') {
+		return false;
+	}
+
+	/* Recursively call GetKey method for the parents until the full key name including all hierarchy levels will be
```

Comment:
It didn't come to my mind. I can see it indeed implements similar relationship to the one I created (actually it links elements in both directions, but nvm), but how it would allow to avoid recursion? I still would need to call GetKey() for each of the elements in hierarchy to concatenate the full key name.

Reply:
Well, like iterating through all elements of the list? I don't mind recursion, to me it's quite elegant solution, but I recall you mentioning that it's not perfect here - nvm.